### ISSUE TITLE
Exception backtrace has user source as location for stdlib methods

### ISSUE THREAD

**Issue by segiddins** [oracle/truffleruby#1414]
I ran some code that boiled down to `ruby -e 'File.join :foo, nil'`.

On MRI 2.5.1, the exception is:

```
Traceback (most recent call last):
	1: from -e:1:in `<main>'
-e:1:in `join': no implicit conversion of Symbol into String (TypeError)
```

In truffleruby 1.0.0-rc5, I get:

```
-e:1:in `coerce_to_failed': Coercion error: :foo.to_str => String failed (TypeError)
	from -e:1:in `execute_coerce_to'
	from -e:1:in `coerce_to'
	from -e:1:in `StringValue'
	from -e:1:in `coerce_to_path'
	from -e:1:in `join'
	from -e:1:in `<main>'
```

Which points to methods that _aren't_ on line 1, which I found confusing when trying to hunt down which piece of my code had an issue (trust me, it was more confusing when the stack trace was longer and there was more than 1 line of code)

**Comment by chrisseaton** [comment]
Hi thanks for the report! This is a deliberate decision, designed to mimic what MRI. MRI doesn't report C files as sources for core methods - it repeats the user location, so we decided to do the same thing. It shows up more because our core being in Ruby means more method activations.

You can turn it off with `-Xbacktraces.hide_core_files=false`.

```
$ ruby -Xbacktraces.hide_core_files=false -e 'File.join :foo, nil'
resource:/truffleruby/core/type.rb:112:in `coerce_to_failed': Coercion error: :foo.to_str => String failed (TypeError)
	from resource:/truffleruby/core/type.rb:102:in `execute_coerce_to'
	from resource:/truffleruby/core/type.rb:95:in `coerce_to'
	from resource:/truffleruby/core/kernel.rb:162:in `StringValue'
	from resource:/truffleruby/core/type.rb:498:in `coerce_to_path'
	from resource:/truffleruby/core/file.rb:818:in `join'
	from -e:1:in `<main>'
```

We're open to a discussion about whether we're doing the best thing here. Maybe we should hide the core routines entirely? Or show them fully?

Our core comes originally from Rubinius, and this is what they print for comparison

```
$ ruby -e 'File.join :foo, nil'
                      main # Rubinius::Loader at core/loader.rb:851 (+85)
                     evals # Rubinius::Loader at core/loader.rb:634 (+202)
                      eval # Kernel(Rubinius::Loader) at core/kernel.rb:1142 (+299)
          call_on_instance # Rubinius::BlockEnvironment at core/block_environment.rb:147 (+106)
         { } in __script__ # Object at -e:1 (+8)
                      join . File at core/file.rb:759 (+123)
            coerce_to_path . Rubinius::Type at core/type.rb:619 (+49)
               StringValue # Kernel(Module) at core/kernel.rb:240 (+22)
                 coerce_to . Rubinius::Type at core/type.rb:149 (+30)
         execute_coerce_to . Rubinius::Type at core/type.rb:154 (+12)
   to_str (method_missing) # Kernel(Symbol) at core/zed.rb:1408 (+389)
```

**Comment by eregon** [comment]
One small thing easy to fix here is to have the same message as MRI, I think `Coercion error: :foo.to_str => String failed` is rather unusual and most people are more used to the MRI message for this.
@chrisseaton Could you fix the message?

**Comment by segiddins** [comment]
I disagree about the exception message :) knowing _which_ object is unable to be coerced is very helpful in debugging

**Comment by segiddins** [comment]
Another small comment -- I think I'd rather you either hide the stdlib backtrace entries entirely (which is defacto what MRI does, because so much of the stdlib is implemented in C), or have the actual core file printed -- I found it confusing to see a backtrace entry that had a location that contained entirely different code

**Comment by chrisseaton** [comment]
Right we'll try some more options and think about what to make the default.

**Comment by eregon** [comment]
Exposing core files in backtraces is convenient for debugging but unfortunately leads to quite a few incompatibilities, for example for libraries which check `Kernel#caller`.
(Also core files are loaded from the jar, so they are not actual files on the filesystem.)

So I think it might make sense to try hiding multiple core frames at the top of the stack by default.

**Comment by norswap** [comment]
@eregon Should we close this?

The message is now identical to MRI safe for punctuation (they use `TypeError: <explanation>`, we use `TypeError (<explanation>)`.

`Kernel#caller` etc have specs on them and should behave as in MRI.

**Comment by eregon** [comment]
I'm not sure, we should still think what we can do to make it clearer that it's code part of the core library.

MRI 2.7+ started to use Ruby files in the core library, we should check how they look in the backtraces.

**Comment by eregon** [comment]
Here is the format used by MRI trunk:
```
$ ruby -e 'p [].shuffle!(foo: 1)'                             
Traceback (most recent call last):
	1: from -e:1:in `<main>'
<internal:array>:15:in `shuffle!': unknown keyword: :foo (ArgumentError)
```

We should use something similar for our core files.

**Comment by eregon** [comment]
We need to match one of these patterns in Sinatra for core files:
https://github.com/sinatra/sinatra/blob/91477e6dad7c9d8eb1c358e47b9d83b73d626f84/lib/sinatra/base.rb#L1205-L1214

**Comment by eregon** [comment]
On MRI master:
```
$ ruby -e 'tap { puts caller_locations(0) }'
-e:1:in `block in <main>'
<internal:kernel>:90:in `tap'
-e:1:in `<main>'
```

So I think we're ready to use a similar format now, and always show core library files like:
```
$ truffleruby -e 'tap { puts caller_locations(0) }'
-e:1:in `block in <main>'
<internal:core> core/kernel.rb:527:in `tap'
-e:1:in `<main>'
```

**Comment by eregon** [comment]
Fixed by 68754c65bfaed1f55fd854ae8bccb53da26a0684
